File: BOJ_3055.py
#탈출
#bfs

import sys
from collections import deque

input = sys.stdin.readline
r, c = map(int, input().split())

graph = [list(input().strip()) for _ in range(r)]
distance = [[0] *c for _ in range(r)]
dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]
queue = deque()

def bfs(Dx,Dy):
    while queue:
        x,y = queue.popleft()
        if graph[Dx][Dy] == 'S': #비버의집 위치가 S이면 distance 반환
            return distance[Dx][Dy]
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < r and 0 <= ny < c:
                if (graph[nx][ny] == '.' or graph[nx][ny] == 'D') and graph[x][y] == 'S': #현재위치가 S이면 고슴도치가 움직일 차례 (고슴도치는 . 이거나 D일때만 이동가능)
                    graph[nx][ny] = 'S'
                    distance[nx][ny] = distance[x][y] + 1
                    queue.append((nx,ny))
                elif (graph[nx][ny] =='.' or graph[nx][ny] =='S') and graph[x][y] == '*': #현재위치가 * 이면 물이 움직일 차례 (물은 . 와 S일때만 이동 가능)
                    graph[nx][ny] = '*'
                    queue.append((nx,ny))
    return "KAKTUS"


for x in range(r):
    for y in range(c):
        if graph[x][y] == 'S':
            queue.append((x,y))
        elif graph[x][y] == 'D':
            Dx,Dy = x,y

for x in range(r):
    for y in range(c):
        if graph[x][y] == '*':
            queue.append((x,y))

print(bfs(Dx,Dy))

File: BOJ_3055.java
import java.awt.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class BOJ_3055 {
    static int R;
    static int C;
    static char[][] map;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int ans = Integer.MAX_VALUE;
    static Queue<Point> que = new LinkedList<>();
    static final String cantgo = "KAKTUS";
    static int sr,sc,dr,dc;
    static int[][] time;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] rc = br.readLine().split(" ");
        R = Integer.parseInt(rc[0]);
        C = Integer.parseInt(rc[1]);

        map = new char[R][C];
        time = new int[R][C];

        for (int i = 0; i < R; i++) {
            char[] inputStr = br.readLine().toCharArray();
            map[i] = inputStr;
            for (int j = 0; j < C; j++) {
                if (map[i][j] == 'S'){
                    sr = i;
                    sc = j;
                    que.offer(new Point(i, j));
                }
                else if (map[i][j] == 'D'){
                    dr = i;
                    dc = j;
                }
            }
        }

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (map[i][j] == '*'){
                    que.offer(new Point(i, j));
                }
            }
        }

        if(bfs(dr, dc)){
            System.out.println(time[dr][dc]);
        }
        else{
            System.out.println(cantgo);
        }

        br.close();
    }

    private static boolean bfs(int x, int y) {
        while(!que.isEmpty()){
            Point now = que.poll();
            //System.out.println(now.x+": "+now.y);
            if (map[x][y] == 'S'){
                return true;
            }

            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if (inRange(nx, ny)){
                    if ((map[nx][ny] == '.' || map[nx][ny] == 'D') && map[now.x][now.y] == 'S'){
                        map[nx][ny] = 'S';
                        time[nx][ny] = time[now.x][now.y] + 1;
                        que.offer(new Point(nx, ny));
                    }
                    else if ((map[nx][ny] == '.' || map[nx][ny] == 'S') && map[now.x][now.y]  == '*'){
                        map[nx][ny] = '*';
                        que.offer(new Point(nx, ny));
                    }
                }
            }

        }
        return false;
    }

    private static boolean inRange(int nx, int ny) {
        return 0 <= nx && nx < R && ny >= 0 && ny < C;
    }

}
